blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +365,229 @@ where
+    /// before returning.
+    pub async fn wait_for_message(
+        &self,
+        cid: &Cid,
+        confidence: i64,
+    ) -> Result<(Option<Arc<Tipset>>, Option<MessageReceipt>), Error> {
+        let mut subscribers = self.subscriber.clone().ok_or_else(|| {
+            Error::Other("State Manager not subscribed to tipset head changes".to_string())
+        })?;
+        let (sender, receiver) = oneshot::channel::<()>();
+        let message = chain::get_chain_message(self.get_block_store_ref(), cid)
+            .map_err(|err| Error::Other(format!("failed to load message {:}", err)))?;
+
+        let maybe_subscriber: Option<HeadChange> = subscribers.next().await;
+        let first_subscriber = maybe_subscriber.ok_or_else(|| {
+            Error::Other("SubHeadChanges first entry should have been one item".to_string())
+        })?;
+
+        let tipset = match first_subscriber {
+            HeadChange::Current(tipset) => tipset,
+            _ => {
+                return Err(Error::Other(format!(
+                    "expected current head on SHC stream (got {:?})",
+                    first_subscriber
+                )))
+            }
+        };
+        let maybe_message_reciept = self.tipset_executed_message(&tipset, cid, &message)?;
+        if let Some(r) = maybe_message_reciept {
+            return Ok((Some(tipset.clone()), Some(r)));
+        }
+
+        let mut candidate_tipset: Option<Arc<Tipset>> = None;
+        let mut candidate_receipt: Option<MessageReceipt> = None;
+        let mut back_tipset: Option<Tipset> = None;
+        let mut back_receipt: Option<MessageReceipt> = None;
+        async {
+            let (back_t, back_r) = self
+                .search_back_for_message(&tipset, &message)?
+                .ok_or_else(|| {
+                    Error::Other("State manager not subscribed to back search wait".to_string())
+                })?;
+            back_tipset = Some(back_t);
+            back_receipt = Some(back_r);
+            sender
+                .send(())
+                .map_err(|e| Error::Other(format!("Could not send to channel {:?}", e)))?;
+            Ok::<(), Error>(())
+        }
+        .await?;
+
+        let height_of_head = tipset.epoch();
+        let mut reverts: HashMap<TipsetKeys, bool> = HashMap::new();
+        let mut back_search_recieved = None;
+        async {
+            back_search_recieved =
+                Some(receiver.await.map_err(|e| {
+                    Error::Other(format!("Could not receieve from channel {:?}", e))
+                })?);
+            Ok::<(), Error>(())
+        }
+        .await?;
+
+        while let Some(subscriber) = self
+            .subscriber
+            .clone()
+            .ok_or_else(|| {
+                Error::Other("State Manager not subscribed to tipset head changes".to_string())
+            })?
+            .next()
+            .await
+        {
```

Comment:
This should be polled at the same time as the subscriber (both the receiver channel and pubsub receiver are futures) as right now you are waiting for the function to complete, then awaiting on the subscriber events. These changes don't change anything I don't think (unless I'm reading wrong)